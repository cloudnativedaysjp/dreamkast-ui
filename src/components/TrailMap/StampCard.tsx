import React, { useContext, useEffect, useState } from 'react'
import * as Styled from './styled'
import { useSelector } from 'react-redux'
import {
  settingsInitializedSelector,
  settingsSelector,
} from '../../store/settings'
import { pointsSelector, useStamps } from '../../store/points'
import {
  usePostApiV1AppDataByProfileIdConferenceAndConferenceMutation,
  usePostApiV1ProfileByProfileIdPointMutation,
} from '../../generated/dreamkast-api.generated'
import {
  getQRCodeStampResult,
  clearQRCodeStampResult,
  getSessionEventNum,
} from '../../util/sessionstorage/trailMap'
import { EnvCtx } from '../../context/env'
import { Skeleton } from '@material-ui/lab'

type Props = {
  todo?: boolean
}

export const StampCard = (_: Props) => {
  const envCtx = useContext(EnvCtx)
  const settings = useSelector(settingsSelector)
  const points = useSelector(pointsSelector)
  const initialized = useSelector(settingsInitializedSelector)
  const stamp = useStamps()
  const [alreadyAdded, setAlreadyAdded] = useState<boolean>(false)
  const [pinnedStamp, setPinnedStamp] = useState<typeof stamp | null>(null)
  const [stamped, setStamped] = useState<boolean>(false)
  const [mutateAppData] =
    usePostApiV1AppDataByProfileIdConferenceAndConferenceMutation()
  const [postPointEvent] = usePostApiV1ProfileByProfileIdPointMutation()

  useEffect(() => {
    if (!points.appDataInitialized || pinnedStamp !== null) {
      return
    }
    setPinnedStamp(stamp)
  }, [stamp, pinnedStamp])

  // get stamp by offline user via QR code
  useEffect(() => {
    const res = getQRCodeStampResult()
    if (!res) {
      return
    } else if (res === 'ok') {
      setStamped(true)
      setAlreadyAdded(true)
    } else {
      // TODO show error info
      console.error(`unexpected result: ${res}`)
    }
    clearQRCodeStampResult()
  }, [])

  // get stamp by online user
  useEffect(() => {
    if (!initialized) {
      return
    }
    if (!stamp.canGetNewStamp) {
      return
    }
    ;(async () => {
      if (!stamp.slotIdToBeStamped) {
        return
      }
      const eventNum = getSessionEventNum(stamp.slotIdToBeStamped)
      const pointEventId = envCtx.getPointEventId(eventNum)
      try {
        await postPointEvent({
          profileId: `${settings.profile.id}`,
          profilePoint: {
            conference: settings.eventAbbr,
            pointEventId,
          },
        })
        await mutateAppData({
          profileId: `${settings.profile.id}`,
          conference: settings.eventAbbr,
          dkUiDataMutation: {
            action: 'stampedFromUI',
            payload: {
              slotId: stamp.slotIdToBeStamped,
            },
          },
        })
        setStamped(true)
      } catch (err) {
        console.error('stampFromUI Action', err)
      }
    })()
  }, [initialized, stamp.canGetNewStamp])

  if (!pinnedStamp) {
    return (
      <Styled.StampCardContainer>
        <div className={'suspend'}>
          <Skeleton variant="rect" animation={'wave'} className={'skeleton'} />
        </div>
      </Styled.StampCardContainer>
    )
  }

  const StampsNotYetAdded = stampLocation.map((loc, i) => (
    <Styled.StampFrame top={`${loc.top}%`} left={`${loc.left}%`}>
      {i < pinnedStamp.stamps.length ? (
        <Styled.Stamp src={`/cndt2022/ui/cndt2022_stamp.png`}></Styled.Stamp>
      ) : (
        ''
      )}
      {i === pinnedStamp.stamps.length && stamped ? (
        <Styled.Stamp
          className={'showAnimation'}
          src={`/cndt2022/ui/cndt2022_stamp.png`}
        ></Styled.Stamp>
      ) : (
        ''
      )}
    </Styled.StampFrame>
  ))
  const StampsAlreadyAdded = stampLocation.map((loc, i) => (
    <Styled.StampFrame top={`${loc.top}%`} left={`${loc.left}%`}>
      {i < pinnedStamp.stamps.length - 1 ? (
        <Styled.Stamp src={`/cndt2022/ui/cndt2022_stamp.png`}></Styled.Stamp>
      ) : (
        ''
      )}
      {i === pinnedStamp.stamps.length - 1 && stamped ? (
        <Styled.Stamp
          className={'showAnimation'}
          src={`/cndt2022/ui/cndt2022_stamp.png`}
        ></Styled.Stamp>
      ) : (
        ''
      )}
    </Styled.StampFrame>
  ))

  return (
    <>
      <Styled.StampCardContainer>
        <Styled.StampCard src={`/cndt2022/ui/stamp_bg.jpg`}></Styled.StampCard>
        {alreadyAdded ? StampsAlreadyAdded : StampsNotYetAdded}
      </Styled.StampCardContainer>
    </>
  )
}

const stampLocation = [
  {
    top: 32,
    left: 7,
  },
  {
    top: 32,
    left: 22,
  },
  {
    top: 32,
    left: 37,
  },
  {
    top: 32,
    left: 52,
  },
  {
    top: 32,
    left: 67,
  },
  {
    top: 32,
    left: 82,
  },

  {
    top: 70,
    left: 7,
  },
  {
    top: 70,
    left: 22,
  },
  {
    top: 70,
    left: 37,
  },
  {
    top: 70,
    left: 52,
  },
  {
    top: 70,
    left: 67,
  },
  {
    top: 70,
    left: 82,
  },
]
