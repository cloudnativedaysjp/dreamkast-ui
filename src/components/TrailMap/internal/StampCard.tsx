import React, { useEffect, useState } from 'react'
import * as Styled from '../styled'
import { useSelector } from 'react-redux'
import { appDataSelector, useStamps } from '../../../store/appData'
import { Skeleton } from '@material-ui/lab'
import { useAddStampIfSatisfied } from '../hooks'
import { DkUiData } from '../../../generated/dreamkast-api.generated'
import { stampLocation } from './const'
import { useImgPreload } from '../../common/PreloadedImg'

const imgStampBg = `/cicd2023/ui/stamp_bg.jpg`
const imgStamp = `/cicd2023/ui/cicd2023_stamp.png`

export const StampCard = () => {
  const { appDataInitialized } = useSelector(appDataSelector)
  const { stamps } = useStamps()
  const { addedNew, addedByQRCode } = useAddStampIfSatisfied()

  const [pinnedStamps, setPinnedStamps] = useState<
    DkUiData['stampChallenges'] | null
  >(null)

  useEffect(() => {
    if (!appDataInitialized) {
      return
    }
    if (pinnedStamps !== null) {
      return
    }
    setPinnedStamps(stamps)
  }, [stamps, pinnedStamps])

  useImgPreload(imgStampBg)
  useImgPreload(imgStamp)

  if (!pinnedStamps) {
    return (
      <Styled.StampCardContainer>
        <div className={'suspend'}>
          <Skeleton variant="rect" animation={'wave'} className={'skeleton'} />
        </div>
      </Styled.StampCardContainer>
    )
  }

  const showStampWithoutEffect = addedByQRCode
    ? (i: number) => i < pinnedStamps.length - 1
    : (i: number) => i < pinnedStamps.length

  const showStampWithEffect = addedByQRCode
    ? (i: number) => i === pinnedStamps.length - 1 && addedNew
    : (i: number) => i === pinnedStamps.length && addedNew

  return (
    <PStampCard
      showStampWithoutEffect={showStampWithoutEffect}
      showStampWithEffect={showStampWithEffect}
    ></PStampCard>
  )
}

type Props = {
  showStampWithoutEffect: (i: number) => boolean
  showStampWithEffect: (i: number) => boolean
}

export const PStampCard = ({
  showStampWithoutEffect,
  showStampWithEffect,
}: Props) => {
  return (
    <>
      <Styled.StampCardContainer>
        <Styled.StampCard src={imgStampBg}></Styled.StampCard>
        {stampLocation.map((loc, i) => (
          <Styled.StampFrame key={i} top={`${loc.top}%`} left={`${loc.left}%`}>
            {showStampWithoutEffect(i) && (
              <Styled.Stamp src={imgStamp}></Styled.Stamp>
            )}
            {showStampWithEffect(i) && (
              <Styled.Stamp
                className={'showAnimation'}
                src={imgStamp}
              ></Styled.Stamp>
            )}
          </Styled.StampFrame>
        ))}
      </Styled.StampCardContainer>
    </>
  )
}
