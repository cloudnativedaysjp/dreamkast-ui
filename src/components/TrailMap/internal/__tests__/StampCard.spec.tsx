import React from 'react'
import { PStampCard } from '../StampCard'
import { render } from '@testing-library/react'

describe('PStampCard', () => {
  it('should render without crash', () => {
    const screen = render(
      <PStampCard
        showStampWithoutEffect={(i) => [0, 1, 2, 6, 7, 8].includes(i)}
        showStampWithEffect={(i) => i === 9}
      />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })
  it('should render when not stamped at all', () => {
    const screen = render(
      <PStampCard
        showStampWithoutEffect={() => false}
        showStampWithEffect={() => false}
      />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })
  it('should render when all stamped', () => {
    const screen = render(
      <PStampCard
        showStampWithoutEffect={() => true}
        showStampWithEffect={() => false}
      />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })
})
