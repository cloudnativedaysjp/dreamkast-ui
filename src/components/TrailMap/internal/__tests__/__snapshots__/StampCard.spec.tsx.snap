// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PStampCard should render when all stamped 1`] = `
<DocumentFragment>
  <div
    class="styled__StampCardContainer-sc-1t1auio-6 VGgex"
  >
    <img
      class="styled__StampCard-sc-1t1auio-7 jeBhwm"
      src="/cicd2023/ui/stamp_bg.jpg"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 htipaU"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 bATKRm"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 cgNZFZ"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 caRSaT"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 jbvwfG"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 kdXKwo"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PStampCard should render when not stamped at all 1`] = `
<DocumentFragment>
  <div
    class="styled__StampCardContainer-sc-1t1auio-6 VGgex"
  >
    <img
      class="styled__StampCard-sc-1t1auio-7 jeBhwm"
      src="/cicd2023/ui/stamp_bg.jpg"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 htipaU"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 bATKRm"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 cgNZFZ"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 caRSaT"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 jbvwfG"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 kdXKwo"
    />
  </div>
</DocumentFragment>
`;

exports[`PStampCard should render without crash 1`] = `
<DocumentFragment>
  <div
    class="styled__StampCardContainer-sc-1t1auio-6 VGgex"
  >
    <img
      class="styled__StampCard-sc-1t1auio-7 jeBhwm"
      src="/cicd2023/ui/stamp_bg.jpg"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 htipaU"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 bATKRm"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 cgNZFZ"
    >
      <img
        class="styled__Stamp-sc-1t1auio-9 hKSnww"
        src="/cicd2023/ui/cicd2023_stamp.png"
      />
    </div>
    <div
      class="styled__StampFrame-sc-1t1auio-8 caRSaT"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 jbvwfG"
    />
    <div
      class="styled__StampFrame-sc-1t1auio-8 kdXKwo"
    />
  </div>
</DocumentFragment>
`;
